mds/utils.py

```diff
@@ -25,3 +31,71 @@ def get_random_point(polygon):
         point = Point(random.uniform(x_min, x_max), random.uniform(y_min, y_max))
         if polygon.contains(point):
             return point
+
+
+def upsert_event_records(
+    event_records: types.GeneratorType, source: str, on_conflict_update=False
+):
+    """
+    Using "upsert" to create event or telemetry records.
+
+    Records pushed by providers have precedence over the ones we pull from them.
+    So we overwrite any existing record with the same timestamp for a given device.
+
+    Lines will be created or updated with the save time of the transaction.
+
+    Args:
+        event_records: list of EventRecord
+        source: "push" or "pull"
+        on_conflict_update: ignore duplicates (default) or overwrite
+    """
+
+    def serialize(event_record):
+        event_record.full_clean()
```

Comment:
@cedricco alternative where I accept (unsaved) EventRecord instances and validate using the model, rather than reinventing it. Good idea?

Reply:
Seems fair ! the only drawback I see could be performance issues